bpo-41818: Make additions to the tty module

8vasu opened the pull request:
This follows #23536. Also, see #23686, #23740.

Add `cfmakeraw()`, `cfmakeecho()`, and `cfmakecbreak()`; make `setraw()` and `setcbreak()` return saved termios attributes to avoid extra `tcgetattr()` calls.

Post #23686, #23546, #23740 goals:
1. add `test_winsize()` to "Lib/test/test_pty.py";
2. major revision of "Lib/pty.py", which has not happened since the beginning of the millennium.

Notes
---
1. List of all POSIX.1-2017 input mode flags - https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/termios.h.html#tag_13_72_03_02.
2. List of all POSIX.1-2017 local mode flags - https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/termios.h.html#tag_13_72_03_06.
3. Naming justification for the functions: https://man7.org/linux/man-pages/man3/cfmakeraw.3.html Note that `cfmakeraw()` is not POSIX compliant; it is available on the BSDs and in glibc. However, POSIX has functions with similar names; for example, consider `cfgetispeed()`. Please see https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/termios.h.html.
4. `test_winch()` will be resubmitted as a part of a later pull request.
5. #23686 replaces `class winsize`.
6. #23740 replaces `tty.login_tty()`.

Signed-off-by: Soumendra Ganguly <soumendraganguly@gmail.com>

<!-- issue-number: [bpo-41818](https://bugs.python.org/issue41818) -->
https://bugs.python.org/issue41818
<!-- /issue-number -->

asvetlov commented on Lib/tty.py:
```diff
@@ -15,22 +24,81 @@
-    mode[CFLAG] = mode[CFLAG] | CS8
-    mode[LFLAG] = mode[LFLAG] & ~(ECHO | ICANON | IEXTEN | ISIG)
+try:
+    CHECK_WINSZ = TIOCGWINSZ
```
Duplicate

8vasu replied:
As you suggested, I will use `hasattr()` for this instead.

asvetlov commented on Lib/tty.py:
```diff
@@ -15,22 +24,81 @@
+        If fd is not a descriptor of
+        a tty, then OSError is raised."""
+        ioctl(fd, TIOCSWINSZ, winsz)
+except NameError:
```
Explicit check for `hasattr(termios, "TIOCSWINSZ")` or something like this is better than `except NameError`.

8vasu replied:
It did feel like there had to be a better way of doing this. I did not like the `except NameError` either. Thank you so much for this suggestion.

8vasu replied:
I have one concern. To be able to use `hasattr(termios, "TIOCSWINSZ")`, I think I need to `import termios`. Am I correct about this? Since we already have `from termios import *`, which cannot be removed because of backward compatibility, is it good practice to also have an `import termios` there just for that single purpose? I was thinking that we could use

```
try:
   from termios import TIOCGWINSZ, TIOCSWINSZ
   ...
except ImportError:
   ...
```

which seems to be an explicit check as well.

asvetlov commented on Lib/tty.py:
```diff
@@ -15,22 +24,81 @@
+        ioctl(fd, TIOCSWINSZ, winsz)
+except NameError:
+    HAVE_WINSZ = False
+    getwinsz = None
```
I think better to make a stub that raises `NotImplementedError` is better than setting the function name to `None`.
In std lib we never use this pattern.

Sometimes we don't declare unsupported function at all, but in this case we usually mark in docs: "Availability: Windows" or "Availability: Linux kernel >= 4.5 or glibc >= 2.27".

I think here is better to support stubs that always fail.

8vasu replied:
I am working on this.

asvetlov commented on Lib/tty.py:
```diff
@@ -15,22 +24,81 @@
+    getwinsz = None
+    setwinsz = None
+
+def mode_echo(mode, echo=True):
```
`mode_echo` is not the best name I guess.
Looking on the function body I can suggest renaming with `enable_echo()`.

`enable_echo(mode)` and `enable_echo(mode, False)`.
Maybe others have better idea?

8vasu replied:
`enable_echo()` does sound better than `mode_echo()`. If you agree with `makeraw()` instead of `mode_raw()`, then we could also change this to `makeecho()`; however, I think that `enable_echo()` sounds better than either. I am leaning toward `enable_raw()`, and the updated commit will use that name unless someone else suggests something better.

8vasu replied:
How about `mkecho()`? See next section.

asvetlov replied:
Sounds good

asvetlov commented on Lib/tty.py:
```diff
@@ -15,22 +24,81 @@
+    else:
+        mode[LFLAG] &= ~ECHO
+
+def mode_raw(mode):
```
`enable_raw()` maybe if we agree with `enable_echo()`?

8vasu replied:
Yes, that was a temporary name. This is supposed to function like `cfmakeraw(3)`. While `enable_raw()` does sound better than `mode_raw()`, we could also change it to `tty.makeraw()` if you permit me to do so, and if there are no licensing issues.

8vasu replied:
How about `mkraw()` instead of `makeraw()`? Reasoning: `mkecho()` seems better than `makeecho()`.

asvetlov replied:
Ok

8vasu replied:
I have changed the names to `cfmakeraw()`, `cfmakeecho()`, ... so that they conform to standard naming convention.

asvetlov commented on Lib/tty.py:
```diff
@@ -15,22 +24,81 @@
+    tcsetattr(fd, when, new)
+    return mode
+
+def login(fd):
```
The naming again.
For me, the function does not *login* but *makes the calling process a session leader* as you correctly wrote in docstring.

Maybe `make_session_leader()` or `set_session_leader()` is better?

8vasu replied:
Perfect! I will use one of those, or a variant of one of those. I agree that `tty.login()` was poor choice on my part; I wanted it to sound like `login_tty(3)`.

8vasu replied:
Update: since I had written these a while ago, I had forgotten what this function exactly does. I did some research, and it turns out that this function is actually doing a bit more than making the calling process a session leader.

According to https://pubs.opengroup.org/onlinepubs/9699919799/functions/setsid.html, the `setsid()` call in our function is responsible for making the calling process a session leader; moreover

> Upon return the calling process shall be the session leader of this new session, shall be the process group leader of a new process group, and **shall have no controlling terminal**.

Our function is also making `fd` the stdin/stdout/stderr and **controlling tty** of the calling process, which is why it belongs in the tty module.

Nevertheless, as you said, `login()` sounds like a bad choice; I completely agree with that. Should I change it to `login_tty()`? This function IS actually a clone of `login_tty(3)`.

Additional note: on my Debian system, `man login_tty` says

> These  are BSD functions, present in glibc. They are not standardized in POSIX.

asvetlov replied:
Maybe we should add `pty.login_tty()` function instead which calls the native `login_tty(3)`?
It can do some platform-specific things which we don't want to learn and reimplement in Python code.

8vasu replied:
Sir, this sounds like a great idea. I have removed `tty.login_tty()` from this pull request. I have also done restructuring/cleanup. Please check the comments on the diff.

8vasu replied:
Please see #23740 for my implementation of `os.login_tty()`.

asvetlov reviewed (changes requested):
I suggest esablish public function names first, update the `tty` module documentation, and make the final review.

@jonathanslenders I very respect your work on https://github.com/prompt-toolkit/python-prompt-toolkit 
Maybe you want to review this (and follow-up) PR? Maybe you can suggest something? I really like to hear voices from experts in PTY/TTY area.

jonathanslenders commented on Lib/tty.py:
```diff
@@ -15,22 +24,93 @@
+    try:
+        ioctl(STDOUT_FILENO, TIOCSCTTY)
+    except (NameError, OSError):
+        tmp_fd = os.open(os.ttyname(STDOUT_FILENO), os.O_RDWR)
```
I'm not sure what this line is doing exactly? Maybe add a comment here?

8vasu replied:
Here is the source of `login_tty(3)`: https://sourceware.org/git/?p=glibc.git;a=blob;f=login/login_tty.c

The
```
    except (NameError, OSError):
        tmp_fd = os.open(os.ttyname(STDOUT_FILENO), os.O_RDWR)
```
in our code is equivalent to their

```
 {
        /* This might work.  */
        char *fdname = ttyname (fd);
        int newfd;
        if (fdname)
          {
            if (fd != 0)
              (void) close (0);
            if (fd != 1)
              (void) close (1);
            if (fd != 2)
              (void) close (2);
            newfd = open (fdname, O_RDWR);
            (void) close (newfd);
          }
 }
```

I contacted the glibc people a while ago and this is what they had to say

> Since this was added by Roland almost 25 years ago, it seems to be trying to
emulate the TIOCSCTTY for system derived from System V (as indicated by this
stackflow thread [1]).

> Back in the day glibc tries to enclose fallbacks to be portable to the various
Unix implementations.  Currently this is just a dead code with various
shortcomings (as for a lot of the glibc fallback implementation where proper
kernel support was the correct answer) since neither Linux nor Hurd uses it.

> And it seems that Unixes derived from System V also has moved on from this
interface, neither Solaris nor Illumos seems to provide it.

> https://urldefense.com/v3/__https://stackoverflow.com/questions/51593530/code-explanation-for-glibc-login-tty-function-openttyname-immediately-f__;!!KwNVnqRv!Vtu1n8f09GnQ28wafOTRPvyTWxkgO0bd7c3sHMxR5GQqPjq4jJhaL_8J8kw-ZliJ$

TL;DR: `TIOCSCTTY` is available everywhere now. Therefore, I too think that it would be safe to remove the `except (NameError, OSError):` block. **Should I remove it?**

I am going to use `tty.login_tty()` [ or whatever we end up naming it ] to replace the fallback code in `pty.fork()`, which still uses the old System V approach. `tty.login_tty()` will also be used in `pty.spawn()` to make it easier for us to implement some new/improved functionality.

jonathanslenders commented on Lib/tty.py:
```diff
@@ -15,22 +24,93 @@
+def login_tty(fd):
+    """Makes the calling process a session leader; if fd is a
+    descriptor of a tty, then that tty becomes the stdin, stdout,
+    stderr, and controlling tty of the calling process. Closes fd."""
```
This comment implies that fd can either be or not be a descriptor of a tty.
Not sure what would happen if it's not a tty descriptor.

8vasu replied:
You are right. I will have to rephrase that. Actually, it will fail if `fd` is not a tty. **Do you think it would be a good idea to instead mention/document what exception would be raised if fd was not a tty?**

8vasu committed:
Soumendra Ganguly: new additions to the tty library. Functions added: cfmakeecho(), cfmakeraw(), and cfmakecbreak().
The functions setcbreak() and setraw() now return original termios to save an extra tcgetattr() call.

Signed-off-by: Soumendra Ganguly <soumendraganguly@gmail.com>

8vasu committed:
Clear all POSIX.1-2017 local mode flags in tty.cfmakeraw().

Signed-off-by: Soumendra Ganguly <soumendraganguly@gmail.com>

8vasu committed:
Remove news file.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.
